### PR TITLE
Reworked build script to build multiple versions of each package

### PR DESCRIPTION
**Pull request by ydesgagn**
Improvements:
1. Added more automation to install toolchain and dependencies.
2. Added a loop to build all Python versions, including 3.11 all at once.
3. Versions are retrieved for each package, and we try building every tag with no letter after 2021.

```
# first time will build deps and all packages with a tag higher than 2021
./make.sh

# debug a package after dependencies and environment are installed
./make.sh --skip-deps --skip-envs --package package_name

# build less versions
./make.sh --year 2022
./make.sh --skip-deps --skip-envs --year 2022 --package package_name
```

Issues:
1. The build script edits the version in each yaml file so it does appear as a diff to git. Maybe best to pass it as a parameter to `build-wheel.py`?
2. Some packages like preshed are depending on Cython. How do you build that?
3. I did not change any versions in the yaml and some packages cannot apply patch files like murmurhash or statsmodels. Something not committed or they were not building?
4. `make OpenSSL-wheels` doesn't work for me in apple support so I cannot build cryptography. All others work fine.
5. twisted doesn't build with `AttributeError: twisted has no attribute __version__`.

To Discuss:
1. I added a file `versions.py` but we may consider moving one loop from `make.sh` doing all versions directly into `build-wheel.py` to keep all the Python code together?

### CODE REVIEW

File: server/pypi/make.sh
Comment (freakboy3742) [inline server/pypi/make.sh]: FYI: Yes, we should; they're currently disabled because the configurations don't build successfully.

File: server/pypi/make.sh
Comment (freakboy3742) [inline server/pypi/make.sh]: Is there anything specific that we're getting out of using Conda here? Is it purely to establish that we have a specific Python version (and can iterate through Python versions)?

If that's the case, my preference would be to make the environment creation portions an external pre-requisite to this script - i.e., make the script assume that it will be run inside a valid Python3.X environment. That (a) simplifies this script, and (b) means the script can be run in parallel, once for each Python version, with independent success/failure on each Python version. It also means we don't have to enforce a dependency on Conda.

File: server/pypi/make.sh
Comment (freakboy3742) [inline server/pypi/make.sh]: Is this a TODO for a reason? I'm not aware of any issues building OpenSSL-wheels

File: server/pypi/make.sh
Comment (freakboy3742) [inline server/pypi/make.sh]: As with the packages- these are dependencies that didn't build for some reason. The only one I can see that *isn't* needed is libffi, but only because that's built as part of Apple-Support.

File: server/pypi/make.sh
Comment (freakboy3742) [inline server/pypi/make.sh]: It took me a moment to work out why the hard-code version of Python; once I worked out that it was for dependencies only, it made more sense. However, the 3.10 version will need to be updated at some point in the future, and will need to be updated in multiple places because of being hard coded. 

Also - following an earlier comment - the hard-coding becomes unnecessary if the script is made "single Python version", as the dependencies pass can be done using whatever Python is in the build environment. In terms of parallelisation, that means we would only need to do a dependencies pass on *one* Python version, then do all the other Python versions (or do "just dependencies" on on Python version, and then "packages" on all versions)

File: server/pypi/make.sh
Comment (freakboy3742) [inline server/pypi/make.sh:97]: This is a neat trick! 

Does this require any additional handling to account for patches that drift over time? I'd be surprised if the same iOS patches apply cleanly for every release over a multi-year window. What's the failure mode when this sort of drift occurs?

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Thanks - for the most part, this looks like some solid improvements. 

I've left a couple of clarifications and questions inline; but the big question is the meta-level one, about whether the Python environment is something that should be managed *by* this script, or should be a pre-requisite for running the script. My inclination is that it should be a pre-requisite, but I'm interested in your thoughts on that matter.

File: server/pypi/make.sh
Comment (ydesgagn) [reply server/pypi/make.sh]: Will keep that in mind and try to make them all work. Don't wait on this to merge. My goal is to merge something weekly with improved functions/features.

File: server/pypi/make.sh
Comment (ydesgagn) [reply server/pypi/make.sh]: Definitively inside a GitHub action this would be a matrix running the 4 Python versions in parallel without a conda environment.

I'm just trying to have a fixed environment here so we can all work on the same base to fix issues. It doesn't have to be conda at all but it's like easy for us to put each machine in the exact same configuration for now until all packages can build.

I could move it to a second script but the main script must still keep knowledge of it if we keep the loop on the Python version.

I would wait for the harder steps to be completed and take some steps back before we tamper too much with things that may change. YOu said you wanted to keep it common with the main Android repo so we also need to consider this in the mix.

**Comment by freakboy3742** [comment]
Regarding your specific questions:

> 1. The build script edits the version in each yaml file so it does appear as a diff to git. Maybe best to pass it as a parameter to `build-wheel.py`?

Agreed - at the very least, manually modifying the script would make parallelisation near impossible.

> 2. Some packages like preshed are depending on Cython. How do you build that?

Cython is a requirement of the *build* environment, not the runtime environment - so, your local macOS Python environment needs to have Cython installed; but that Cython install generates C code that can be built on iOS. There's no dependency on Cython at runtime; it's a quirk of Python's "build == runtime" assumption leads to Cython being listed as a dependency.

> 3. I did not change any versions in the yaml and some packages cannot apply patch files like murmurhash or statsmodels. Something not committed or they were not building?

STATUS-iOS.md has the current build status, and a summary of the origin of build problems; statsmodels was a known to have issues related to Cython, and something to do with PEP517 running setup.py during download. I'm not aware of any issues with murmurhash, though, and I haven't got anything un-committed or un-pushed in my local repo.

> 4. `make OpenSSL-wheels` doesn't work for me in apple support so I cannot build cryptography. All others work fine.

Hrm... that's weird. Can you share the error you're seeing?

> 5. twisted doesn't build with `AttributeError: twisted has no attribute __version__`.

Again, I don't have any local changes that aren't pushed. However, Twisted uses calver, and the version we've published is 19.10; if you're building versions from 21.0+, then it's possible that they've made a change altering how they get the Twisted version.

> To Discuss:
> 
> 1. I added a file `versions.py` but we may consider moving one loop from `make.sh` doing all versions directly into `build-wheel.py` to keep all the Python code together?

I don't mind there being an external Python file for this sort of functionality (although maybe the script should be in the `env` folder so it's treated as a "build environment" script). While we *could* modify the build-wheel script to build multiple versions, my inclination (as indicated in the review comments) would be to go the other direction - keep build-wheel.py as a "build one wheel" script, and modify this build script to be a "build one Python version" script.

File: server/pypi/make.sh
Comment (ydesgagn) [reply server/pypi/make.sh]: It does build but the wheel doesn't complete.
`
cp -r /Users/yves/Downloads/chaquopy/server/pypi/support/Python-Apple-support/install/iOS/iphonesimulator.x86_64/openssl-3.0.5/include /Users/yves/Downloads/chaquopy/server/pypi/support/Python-Apple-support/install/iOS/iphonesimulator.x86_64/openssl-3.0.5/wheel/opt/include
cp: /Users/yves/Downloads/chaquopy/server/pypi/support/Python-Apple-support/install/iOS/iphonesimulator.x86_64/openssl-3.0.5/include: No such file or directory
make: *** [wheels/dist/openssl/openssl-3.0.5-1-py3-none-ios_12_0_iphonesimulator_x86_64.whl] Error 1
`

File: server/pypi/make.sh
Comment (freakboy3742) [reply server/pypi/make.sh]: Absolutely - multiple small steps is preferable to a monster patch that tries to fix everything in one pass.

File: server/pypi/make.sh
Comment (ydesgagn) [reply server/pypi/make.sh:97]: It's a temp solution. My proposal in the PR text is to merge this with the `build-wheel.py`. We could then only build a default version or many like I do here but yes, to do a nice job those patches should be per version if a match is detected or fallback to a default patch. Honestly, so far very few patches failed. Most of the failures are with Python 3.11.

The patches failing are failing with the unmodified code so we need to look at them.

File: server/pypi/make.sh
Comment (freakboy3742) [reply server/pypi/make.sh]: Having a stable base definitely makes sense (and conda makes at least some sense as a base); however, I'd argue there's a difference between agreeing on a stable base, and baking that in as a requirements of the script's operation. At the very least, I'm a little wary of introducing a hard requirement on Anaconda because of the conflict of interest I have as a result of being employed by Anaconda. An external requirement of "we're going to use Conda for CI purposes, but it will run in any Python environment" is different to an internal requirement of "this script requires that you have conda installed to run it at all".

Your comments you seem to imply that this script could be executed in parallel as-is - I'm not sure I'm see how. What would be the mechanism for this?

Comment (ydesgagn) [reply server/pypi/make.sh]: That script would need serious rewriting to be used inside a parallel CI pipeline. There are two things to consider.

Initial build of a package is probably done manually and will loop on many package versions. This could also be automated in CI with smart commit tags. We could do as part of a commit something like `#build-all-versions` and the loop will be on the CI side. But we need to monitor the CI for failure and some part may fail just because Python 3.11 is not well supported yet and a human will quickly ignore those failures.

Once the initial build is done the CI should take over on monitoring new tag, build only the required new part and deliver the wheel directly.

File: server/pypi/make.sh
Comment (freakboy3742) [reply server/pypi/make.sh]: I wonder if it's a Makefile issue with a missing (or incorrect) dependency. I haven't thoroughly tested each Makefile target; I'm normally running a top level `make all`, and then running `make X-wheels` as an afterthought. I've tried running locally (in an admittedly pre-built environment), and I noticed that `make OpenSSL-wheels` triggered a Python build, which would seem to suggest the build chain isn't exactly "clean" - but the build did complete, including creating the wheels. 

Related - one of the items on my todo list is beeware/Python-Apple-support#169 - reworking the Python support package so that each of the dependencies can be built independently. This will likely involve introducing separate repositories for libFFI, BZip2, XZ and OpenSSL so that we can build those dependencies independently of Python (and have independent publication processes as well).

Comment (ydesgagn) [reply server/pypi/make.sh]: Will try to build all first and test again. Was just trying to save time.

**Comment by ydesgagn** [comment]
Twisted is fine if I go back to 2019 and 2020. So something broke in 2021. Sending you the wheels that are fine.

File: server/pypi/make.sh
Comment (freakboy3742) [reply server/pypi/make.sh]: Ok - but if the end goal is a CI configuration that can build a single version for a single Python, wouldn't that support the idea of having an external "build everything" script that does this sort of iteration, but keep the core build script single target with an externally assumed environment?

Comment (ydesgagn) [reply server/pypi/make.sh]: Yes we agree on this 100%. To me this is `make.sh`. The GitHub action will not need `make.sh` and will directly call `build-wheel.py`.

Let's try to solve a few issues. How do we remove the version from the yaml so it's not linked to a fixed version while at the same time we keep a patch scheme flexible?

I will put in place a prototype a a CI action monitoring remote repo so you better understand.

**Comment by ydesgagn** [comment]
Openssl is also fine if I build everything first. So calling `make OpenSSL-wheels` clearly has a dependency problem. Like you said, it's kind of painful to build the entire Python package just to get tiny dependencies. In the context of a CI environment, we don't want to wait on this (even if it could be cached). The best option is to have binary available in the same or another repo.

**Comment by ydesgagn** [comment]
About the wheels and the dependencies issue, I think we should trust the original package dependencies (automated with queries like I do for the version) for each version. So before building, I could replace the minimum version for each dependency in the yaml but this must clearly be done in `build-wheel.py` when receiving a new version. The package version should be removed from the yaml as well and passed as a parameter.

The Android guys should have the same issue on their side as well if they build multiple versions.

That will probably fail some Python 3.11 wheels where the dependencies cannot be built for some older version but this is not an issue to me if we lose some.

**Comment by freakboy3742** [comment]
> The best option is to have binary available in the same or another repo.

Absolutely agreed. The current setup is something that has evolved over a long time, and is badly in need of a refactor. 

> About the wheels and the dependencies issue, I think we should trust the original package dependencies (automated with queries like I do for the version) for each version. 

To be clear - by "original" do you mean the ones published by the packages themselves (e.g., in numpy sources), or "original" in the sense of the original chaquopy code? I agree that the former would be preferable; the chaquopy definitions only exist because of the complication of getting the build system to install iOS versions of packages on a host build.

> So before building, I could replace the minimum version for each dependency in the yaml but this must clearly be done in `build-wheel.py` when receiving a new version. The package version should be removed from the yaml as well and passed as a parameter.

There are two other options: 
1. We can use meta-packages like oldest-supported-numpy
2. The YAML could be modified to be version-dependent - i.e., specify different dependencies for each Python version.

The latter would be my preferred option, as it's a more general solution, and has a direct analog with normal Python dependencies. Whether we use the `; python_version==3.10` syntax or something else is open to discussion.

> The Android guys should have the same issue on their side as well if they build multiple versions.

Yeah - I discussed this with @mhsmith yesterday; there's definitely a general issue that needs to be addressed here.

> That will probably fail some Python 3.11 wheels where the dependencies cannot be built for some older version but this is not an issue to me if we lose some.

Agreed - I don't think it's reasonable to expect us to maintain support a 3 year old version of a package on a 4 month old version of Python.

**Comment by ydesgagn** [comment]
All comments resolved. This is ready to merge.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This has implemented all the "separation of make-all from make" changes; but it doesn't appear to have addressed the "minimum API version" problem.

I can see you've updated the host numpy versions for a bunch of packages; but (a) it's not clear to me why those specific versions have been selected (since it's not consistent between packages), and (b) it doesn't address the issue of numpy dependency versions being Python version dependent. 

In [this comment](https://github.com/freakboy3742/chaquopy/pull/4#issuecomment-1428840057) (and others in that thread), we discussed a bunch of options for specifying version dependencies; it's not clear if/how this patch is addressing that discussion.

Unless I'm missing something, this set of configurations would require us to build Numpy 1.14 for Python 3.11 in order to publish a current version of shapely; and numpy 1.27 (which doesn't exist?) in order to build scipy. It also means that (based on the most recent archive you've sent me) gensim hasn't been built for Py3.10+ - presumably at least in part because Numpy 1.18.5 isn't available. However, gensim *can* be built for at least Python 3.10 - there's an existing package on the BeeWare repo. Similarly, Pandas doesn't appear to have been built for Python 3.10+, presumably related to the Numpy 1.20.3 requirement.

**Comment by ydesgagn** [comment]
I tried various approaches and none of them were perfect.

Getting the requirements from https://pypi.org/pypi json was inconclusive. Sometimes the have a single line to cover all versions and this doesn't work for Python 3.10 and/or 3.11. It's yielding the same result you evaluated before. In other cases two lines match the Python version (see pandas).

I tried producing a new requirements.in and run pip-compile (from pip-tools). This was fine for regular cases but when you ask for an old version in a Python 3.10 or 3.11 environment pip-compile return and error and doesn't produce a requirmeent.txt file.

The meta.yaml file is not super flexible as it calls for a fixed version and doesn't understand >=.

I came to the conclusion that it's best not to trust any of the sources above and just check the smaller version of the dependency that is available for the current Python version. I did not see any cases where this produces a version number smaller than the minimum requirement in the not trustable package data above but I did not put any protection as the source is not reliable.

This produced more wheels for gensim, matplotlib, pandas and pywavelets for Python 3.10 and 3.11.

Will update the wheels when pandas will be completed in a few hours.

**Comment by mhsmith** [comment]
> The meta.yaml file is not super flexible as it calls for a fixed version and doesn't understand >=.

As currently implemented by the build-wheel tool, the meta.yaml file doesn't support the `; python_version==3.10` syntax mentioned above, or the [conda-build selector syntax](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#preprocessing-selectors). But it does implement Jinja syntax such as `{% if ...`, which could be used to achieve the same effect if the relevant variables were passed into the Jinja renderer.

**Comment by ydesgagn** [comment]
That data is not reliable. For pandas and Python 3.10 I get:

`
numpy (>=1.20.3) ; python_version < "3.10"
numpy (>=1.21.0) ; python_version >= "3.10"
numpy (>=1.23.2) ; python_version >= "3.11"
`

Two lines would match in this case. The two wheels exist for 3.10 so this is fine.

But for gensim this data is the same for all Python versions:

`
numpy (>=1.18.5)
`

The wheel for numpy doesn't exist for 1.18.5. The first available is 1.21.0 for 3.10 and 1.22.0 for 3.11.

But >= support would work in that case but how do you propose the script find the best match? We don't want to redo the pip dependency management and, on top of that just for iOS or Android where not all versions build for those platforms.

I'm for removing the version from the meta.yaml and make is a mandatory parameter instead. We can name the dependencies in the yaml but let the script guess the best version to use like we do now. This would remove the sed and the edditting of git tracked files.

**Comment by freakboy3742** [comment]
> That data is not reliable. For pandas and Python 3.10 I get:
> 
> `numpy (>=1.20.3) ; python_version < "3.10" numpy (>=1.21.0) ; python_version >= "3.10" numpy (>=1.23.2) ; python_version >= "3.11"`
> 
> Two lines would match in this case. The two wheels exist for 3.10 so this is fine.

I'm not sure I see the ambiguity for Python 3.10 (only rule 2 matches); but Python 3.11 matches 2 rules. However, there clearly is *an* interpretation, because Pandas is currently using this build configuration. 

However, they're not version specifiers - they're constraints. The pip/conda solvers take these constraints, and try to find a solution that satisfies them; in this case, for Python 3.11, any version of numpy after 1.21.0 will satisfy the provided constraints. However, in practice, wheels are preferred (unless `--no-binary` or similar directives are employed); a Py3.11 wheel is only available for 1.23.2, so that is the version that will be used.
 
> But for gensim this data is the same for all Python versions:
> 
> `numpy (>=1.18.5)`
> 
> The wheel for numpy doesn't exist for 1.18.5. The first available is 1.21.0 for 3.10 and 1.22.0 for 3.11.

Again it's a constraint, not a specifier. The solution to that constraint is different for different Python versions.

> But >= support would work in that case but how do you propose the script find the best match? We don't want to redo the pip dependency management and, on top of that just for iOS or Android where not all versions build for those platforms.

True, I'd rather not re-implement the full pip solver. However, we don't need a full pip solver - we can limit ourselves to the packages literally described in meta.yaml, and the versions for which wheels are known to exist for a given Python version that also match any Python version-based constraint using `; python_version ...` definitions.
 
> I'm for removing the version from the meta.yaml and make is a mandatory parameter instead. We can name the dependencies in the yaml but let the script guess the best version to use like we do now. This would remove the sed and the edditting of git tracked files.

So - I'm not 100% sure if I agree here. We are defining a build system; that build system should be reproducible. Part of that reproducibility is specifying what inputs were given to the system at the time of build. If the end target is a CI system, a mandatory argument isn't a viable solution - that argument would need to be defined *somewhere*; meta.yaml is as good a place as any.

If you're saying that we can/should completely derive dependency version information based on the published wheels available on PyPI - that's an interesting idea. "Package the app with the versions that were contemperaneous" sounds plausible as an idea, and it certainly appear to be true for Numpy, but I'd want to do some more validation for packages other than Numpy/Pandas.

As a meta-issue - you appear to be continuing to push updates to this PR that are unrelated to this initial "multiple python versions" fix. I haven't done a deep dive on those changes yet, but they seem to be addressing fundamentally different issues. As such, it would be preferable for these to be a second PR (building on the first) so they can be reviewed independently.

**Comment by ydesgagn** [comment]
I can move openblas and the fortran stuff to another PR.

I'm not sure where you guys want to go with the PR. I'm not planning on redesigning your entire system with this PR. I'm just trying to contribute a few more wheels for iOS.

**Comment by freakboy3742** [comment]
> I can move openblas and the fortran stuff to another PR.
> 
> I'm not sure where you guys want to go with the PR. I'm not planning on redesigning your entire system with this PR. I'm just trying to contribute a few more wheels for iOS.

Understood - however, anything that is merged needs to make sense in itself, and be broadly consistent with any future directions we're intending to take. 

To that end: the manual overwriting of meta.yaml doesn't fill me with delight. I can see 3 problems:

1. it's a bit fragile in the general case; 
2. A source checkout is dirtied by running the script. 
3. The end-goal of having the meta.yaml being a stable descriptor of the "current" build state isn't really being honoured, as it doesn't reflect the Python-version-dependent nature of dependencies.

Making the properties that are overwritten something that can be specified as command-line arguments would be my preferred fix for (1) and (2); (3) can't really be fixed without adding some sort of python-version dependency syntax into the interpretation of meta.yaml. 

If you'd rather deal with those problems a follow up PR, I can live with what is in the PR at the moment as a means to an end in the short term, as long as there's some additional "FIXME" style documentation around the sed invocations to make it clear that part of the code is a short term workaround.

The bit I'm not happy with at the moment is the minimum version identification - or, at least, how the current script has been used to generate the packages that are in the archives you've provided. As I understand it, the current script is doing the following:

1. Look at PyPI, look for all final (i.e., non alpha etc_) releases after 2019 (at year-level granularity)
2. Try to build that package for a given Python version
3. If the build succeeds, ship it

Then, for any other package, if numpy is listed as a specific dependency, replace the version pin with the oldest version that the script built successfully.

This seems to generally work, but it does lead to some interesting edge cases - there are a number of packages that are in the archive that support versions of Python that the upstream packages didn't support. For example, there are numpy 1.16.0 packages for Python 3.8 - numpy 1.16.0 was released in Jan 2019, but Python3.8 wasn't even released in *alpha* until Feb 2019. There are similar issues for many packages in the archive you've provided - we're offering a package for versions of Python that didn't exist at the time the package was published. 

I don't think it's reasonable to pretend to offer binary support for versions of Python that the packages themselves didn't pretend to support. It think there's an easy fix for this, though - pass the Python version into the versions script, and use that as part of the filtering criteria. If there isn't a macOS wheel for that version of Python associated with a release, then we don't try to build an iOS wheel for that release.

Here's an updated version of the `versions.py` script that implements this change; it also corrects an issue with root certificates I was seeing in my local installation (I've provided this as a full file rather than trying to create a pull request to your pull request, or pushing to your branch).

```
#!/usr/bin/env python3

import json
import sys
import ssl
from urllib.request import urlopen

import certifi


def get_versions(py_version, package_name, year):
    url = f"https://pypi.org/pypi/{package_name}/json"

    # ensure we're using a root certificate that works with PyPI
    context = ssl.create_default_context(cafile=certifi.where())
    releases = json.load(urlopen(url, context=context))['releases']

    versions = set()

    for version, release in releases.items():
        for package in release:
            if (
                package['packagetype'] == 'bdist_wheel'
                and "-macosx_" in package['filename']
                and int(package['upload_time'].split('-')[0]) >= year
                and not any(c.isalpha() for c in version)
                and package['python_version'] == f"cp{py_version.replace('.', '')}"
            ):
                versions.add(version)

    return sorted(versions)

if __name__ == "__main__":
    year = int(sys.argv[1])
    package_name = sys.argv[2]
    python_version = sys.argv[3]

    print(" ".join(get_versions(python_version, package_name, year)))
```

**Comment by ydesgagn** [comment]
The version was moved out of the yaml files and added as a parameter. I keep the version for the dependencies.

Updated the 'versions.py' with your changes. I even added some command line arguments.

Wheels are uploaded at the same place and it did remove some that were not available on macOS.

FIXME comment added next to the last `sed` for the dependencies.

I'm not too sure what to do with some yaml files like `pyzbar` or `torch` with constructs like this:

```
{% set version = "0.1.8" %}

package:
  name: pyzbar
  version: {{ version }}

source:
  url: https://github.com/NaturalHistoryMuseum/pyzbar/archive/v{{ version }}.tar.gz
```

Also after I took out the version we have some cases like `bcrypt` where we have a build number floating on the air a bit:

```
package:
  name: bcrypt

build:
  number: 2
```

Should we just remove all build number when no version is set?

**Comment by freakboy3742** [comment]
> The version was moved out of the yaml files and added as a parameter. I keep the version for the dependencies.

I'm not sure I understand where this change comes from. Having a command line argument to override default version specified by the configuration file makes sense, but removing the version from the meta.yaml means we don't know what versions in the configuration file actually refer to. The configuration file is (and AFAICT, will continue to be in the "final CI state") the build configuration for the most recent successful build.

> Wheels are uploaded at the same place and it did remove some that were not available on macOS.

I'll do another testing pass; assuming things work, I'll start uploading to the BeeWare repo.
 
> FIXME comment added next to the last `sed` for the dependencies.

> I'm not too sure what to do with some yaml files like `pyzbar` or `torch` with constructs like this:

You shouldn't need to do anything; again, the files that are committed to the repo are the "most recent known good" versions of packages; so the existing configurations shouldn't need any modifications.

(EDIT: After an in-person discussion - the issue is that the *version* will be overwritten, but the other places the templated version is used *won't* be updated - so, in the pyzbar case, the same source URL will be used for every release. We need to modify the template before it is rendered.)
 
> Also after I took out the version we have some cases like `bcrypt` where we have a build number floating on the air a bit:
> 
> ```
> package:
>   name: bcrypt
> 
> build:
>   number: 2
> ```
> 
> Should we just remove all build number when no version is set?

If the version has been overridden at the command line, then the build number should be ignored; but it might be useful to be able to manually set a build number as well.

**Comment by freakboy3742** [comment]
As discussed in person, I wrote up the changes to do the search-and-replace template parsing for version numbers; I've also added in build number handling (purging old build numbers if there's an override version, and allowing for a manually specified build number).

I tried pushing to your branch, but because `iOS-support` is your main branch, I don't have push permissions. Here's the diff for the change; or, if you want to give me push permissions, I'll push from my local version.

[0001-Rewrite-meta.yaml-template-to-override-version-and-b.patch](https://github.com/freakboy3742/chaquopy/files/10789278/0001-Rewrite-meta.yaml-template-to-override-version-and-b.patch)

**Review by freakboy3742 — APPROVED** [review]
I'm happy with where this is now; unless you're expecting additional changes, I'm happy to merge as is.

**Pull request merged**
